cpp/src/io/parquet/page_enc.cu

```diff
@@ -933,6 +939,28 @@ static __device__ void PlainBoolEncode(page_enc_state_s *s,
   }
 }
 
+/**
+ * @brief Converts a 64-bit timestamp into an int96 timstamp that is returned as an int64 and
+ * uint32. Does not deal with time zones.
+ *
+ * @param timestamp uint64 timestamp to convert
+ * @param converted_type logical data type of the timestamp(TIMESTAMP_MILLIS or TIMESTAMP_MICROS)
+ * @return std::pair<uint64_t,uint32_t> where the uint64 is the number of ticks on the last day
+ * since midnight and the uint32 indicates the number of days from Julian epoch.
+ */
+static __device__ std::pair<int64_t, int32_t> convert_64bit_timestamp_to_int96(
+  int64_t timestamp, uint8_t converted_type)
+{
+  constexpr int64_t JulianEpochOffsetDays = INT64_C(2440588);
+  constexpr int64_t MicroSecondsPerDay    = INT64_C(86400000000);
+  constexpr int64_t MilliSecondsPerDay    = INT64_C(86400000);
+
+  auto divisor = converted_type == TIMESTAMP_MILLIS ? MilliSecondsPerDay : MicroSecondsPerDay;
+  uint32_t julian_days   = static_cast<uint32_t>((timestamp / divisor) + JulianEpochOffsetDays);
+  int64_t last_day_ticks = timestamp % divisor;
```

Comment:
IIRC, the ticks in Spark INT96 timestamps was nanoseconds (at least the reader is making that assumption, maybe because converted_type was not present)

Reply:
Are you talking about the value stored in last_day_ticks here? We need to convert the timestamp that is coming in, which could be micro or milliseconds, but I think you are correct that we're not then converting the modulus result into nanoseconds. We need to add something like `(timestamp % divisor) / (TIMESTAMP_MILLIS ? 1000000 : 1000)` to convert to nanoseconds. This brings up some testing questions. I'll have to shore that up as well to ensure we're testing this.

Reply:
Yes, something like `(timestamp % divisor) * (TIMESTAMP_MILLIS ? 1000000 : 1000)` (but maybe also handling the case where the timestamp unit is nanoseconds)
IIRC, the tests tend to use small timestamp values that are multiples of 1000000, so it survives conversion between units without overflowing or losing precision.

Reply:
The python tests were actually using the row number with a -1000 divisor, which meant the stamps were 0/1000, 1/1000, .... This resulted in all timestamps being 0 for the tests, so just about anything could pass the tests. :)